agera/src/main/java/com/google/android/agera/BaseObservable.java

```diff
@@ -125,24 +125,22 @@ void dispatchUpdate() {
     }
 
     private void add(@NonNull final Updatable updatable, @NonNull final Handler handler) {
-      boolean added = false;
+      int indexToAdd = -1;
       for (int index = 0; index < updatablesAndHandlers.length; index += 2) {
         if (updatablesAndHandlers[index] == updatable) {
           throw new IllegalStateException("Updatable already added, cannot add.");
         }
-        if (updatablesAndHandlers[index] == null && !added) {
-          updatablesAndHandlers[index] = updatable;
-          updatablesAndHandlers[index + 1] = handler;
-          added = true;
+        if (updatablesAndHandlers[index] == null && indexToAdd == -1) {
```

Comment:
please think about do `updatablesAndHandlers` shrink in `remove(...)`, for example when `size` become 4/8 times smaller then size of the `updatablesAndHandlers`.